pkgs/native_assets_cli/lib/src/config.dart

```diff
@@ -0,0 +1,538 @@
+
+  /// The asset types that the invoker of this hook supports.
+  final List<String> supportedAssetTypes;
+
+  HookConfig(this.json)
+      : version = switch (Version.parse(json.string(_versionKey))) {
+          final Version version => (version.major != currentVersion.major ||
+                  version < currentVersion)
+              ? throw FormatException(
+                  'Only compatible versions with $currentVersion are supported '
+                  '(was: $version).')
+              : version,
+        },
+        outputDirectory = json.path(_outDirConfigKey),
+        outputDirectoryShared = json.path(_outDirSharedConfigKey),
+        packageRoot = json.path(_packageRootConfigKey),
+        packageName = json.string(_packageNameConfigKey),
+        targetOS = OS.fromString(json.string(_targetOSConfigKey)),
+        supportedAssetTypes =
+            json.optionalStringList(_supportedAssetTypesKey) ?? const [],
+        _buildMode = switch (json.optionalString(_buildModeConfigKey)) {
+          String value => BuildMode.fromString(value),
+          null => null,
+        };
+
+  BuildMode get buildMode {
+    if (_buildMode == null) {
+      throw StateError('Build mode should not be accessed in dry-run mode.');
+    }
+    return _buildMode;
+  }
+
+  @override
+  String toString() => const JsonEncoder.withIndent('  ').convert(json);
+
+  // The latest supported config version.
+  static Version currentVersion = Version(3, 5, 0);
+}
+
+sealed class HookConfigBuilder {
+  final Map<String, Object?> json = {
+    'version': HookConfig.currentVersion.toString(),
+  };
+
+  void setupHookConfig({
+    required Uri packageRoot,
+    required String packageName,
+    required OS targetOS,
+    required List<String> supportedAssetTypes,
+    required BuildMode? buildMode,
+  }) {
+    json[_packageNameConfigKey] = packageName;
+    json[_packageRootConfigKey] = packageRoot.toFilePath();
+    json[_targetOSConfigKey] = targetOS.toString();
+    json[_supportedAssetTypesKey] = supportedAssetTypes;
+    if (buildMode != null) {
+      json[_buildModeConfigKey] = buildMode.toString();
+    }
+  }
+
+  /// Constructs a checksum for a [BuildConfig].
+  ///
+  /// This can be used to construct an output directory name specific to the
+  /// [BuildConfig] being built with this [BuildConfigBuilder]. It is therefore
+  /// assumed the output directory has not been set yet.
+  String computeChecksum() {
+    if (json.containsKey(_outDirConfigKey) ||
+        json.containsKey(_outDirSharedConfigKey)) {
+      // The bundling tools would first calculate the checksum, create an output
+      // directory and then call [BuildConfigBuilder.setupBuildRunConfig] &
+      // [LinkConfigBuilder.setupLinkRunConfig].
+      throw StateError('The checksum should be generated before setting '
+          'up the run configuration');
+    }
+    final hash = sha256
+        .convert(const JsonEncoder().fuse(const Utf8Encoder()).convert(json))
```

Comment:
A few things

* Sharing builds is an optimistic thing: A user can always run `flutter clean` or `rm -rf .dart_tool`. So the fact it does something different now doesn't mean it's wrong.
* The code is deterministic, the maps produced here will have the keys always in the same order in them (our maps have guarantees about iteration order being the same as insertion order). So that means caching will actually work just fine with this code (A Dart SDK with this code in it will build first time and skip second time)

But I have considered making a separate change for this: https://github.com/dart-lang/native/issues/1650

Reply:
Let's do this as an orthogonal piece as part of: https://github.com/dart-lang/native/issues/1650